README.md
# \RoussKS\FinancialYear - Financial Year PHP Library

[![Latest Version](https://img.shields.io/github/release/RoussKS/financial-year.svg?style=round-square)](https://github.com/RoussKS/financial-year/releases)
![Build Status](https://github.com/RoussKS/financial-year/actions/workflows/test.yml/badge.svg)
[![GitHub license](https://img.shields.io/github/license/RoussKS/financial-year.svg)](https://github.com/RoussKS/financial-year/blob/master/LICENSE)

### Introduction / Background / Purpose
In my first working environment as a software/web developer, I stumbled upon the need to do different calculations for financial/accounting year purposes.
The business also happened to have a non-standard calendar financial year (types explained below).
Hence, each time we needed to get any sort of report (transactions, conversion rate, sales etc) we used a predefined list provided by our business analyst. 

This library aims to solve this business problem in a consistent manner.

The calculation of week, period, year and end dates of those for the business.

According to [wikipedia](https://en.wikipedia.org/wiki/Fiscal_year)

>A fiscal year (or financial year, or sometimes budget year) is the period used by governments for accounting and budget purposes, which varies between countries. It is also used for financial reporting by business and other organizations.

An organisation financial year can be based on the following 2 methods:
1. Starting on the same date every year.
   - This is the *`calendar`* type for this library.
2. Ending on the same day of the week every year.
   - This is the *`business`* type for this library.
     >The "fiscal year's end" (FYE) is the date that marks the end of the fiscal year. Some companies—such as Cisco Systems[1]—end their fiscal year on the same day of the week each year, e.g. the day that is closest to a particular date (for example, the Friday closest to 31 December). Under such a system, some fiscal years will have 52 weeks and others 53 weeks.
   - A financial year of this type, always has 364 days a year and is divided in 13 periods (each period has 4 weeks, 28 days).
   - The current library will accommodate the 52-53 weeks methodology of financial year, 
     where a 53rd week is added at the end of a financial year in order to cover for missing days from the previous 364 day years.
     This is a business domain decision, hence the library does not calculate when a year should be 52 or 53 weeks.
     It needs to be set by the user.
     
     
Available methods can be followed through the [RoussKS\FinancialYear\AdapterInterface](https://github.com/RoussKS/financial-year/blob/master/src/AdapterInterface.php) until a full-fledged readme is provided.

### Requirements
- PHP Version ^7.1 || ^8.0 ( 7.1 =< PHP Version =< 8.x.x according to [Composer docs version constraints](https://getcomposer.org/doc/articles/versions.md#caret-version-range-) )

### Installation
```console
composer require roussks/financial-year
```

### Basic Use
```php
require_once __DIR__ . '/vendor/autoload.php';

// DateTimeAdapter
// If instantiating with string, it must be of ISO-8601 format 'YYYY-MM-DD'
$startDate = new \DateTime('2019-01-01');

$fy = new \RoussKS\FinancialYear\DateTimeAdapter('calendar', $startDate);

echo $fy->getFyEndDate()->format('Y-m-d'); // 2019-12-31 
```

### Docker images
The library provides a sample Dockerfile to assist in development use if you want to contribute.
This using the official php cli images.
Copy the `Dockerfile.example` file to Dockerfile, uncommenting the required php version.

However, you are free to use any methodology you want for developing updates & bugfixes.

### Run Tests
The library has an extensive test suite to cover most scenarios & negative paths.

Be aware that some configurations for phpunit are different between versions, 
so you might need to copy the `phpunit.xml.dist` file to `phpunit.xml` and make the necessary adjustments.

```shell
./vendor/phpunit/phpunit/phpunit
```

```shell
XDEBUG_MODE=coverage ./vendor/phpunit/phpunit/phpunit
```


```shell
# HTML coverage example
XDEBUG_MODE=coverage ./vendor/phpunit/phpunit/phpunit --coverage-html tests/report
```

### Run Static Analysis
The library uses PHPStan as a static analysis tool with the default level set to 5.

```shell
./vendor/bin/phpstan analyse
```

### Limitations
Unfortunately, the library does not support a start date of 29, 30, 31 of any month for *`calendar`* financial year type.

We do not expect much use for these dates for a calendar type financial year, as they would cause a considerable problem for accounting.

e.g. if a year starts on 31/1, does the first period end on 28/2? And the following period starts 1/3 effectively skipping a month?

If upon library usage, a user has encountered such a real business issue and can provide a mitigation logic, we can work on implementing it. 

_Important_: This is allowed for a *`business`* type financial year.

### Future Plans
- Update Readme or create Wiki page with more examples of all available methods.

- Introduce new extending library for [CarbonAdapter](https://github.com/roussks/financial-year-carbon) to work directly with [Carbon](https://github.com/briannesbitt/carbon) datetime instances.

- Introduce new extending library for [ChronosAdapter](https://github.com/roussks/financial-year-chronos) to work directly with [Chronos](https://github.com/cakephp/chronos) datetime instances.

- Introduce Laravel Package with different drivers for each of the adapters.

### Versioning
The current library will be using [Semantic Versioning](https://semver.org/)

Non-breaking changes will result in a MINOR or a PATCH version update as classified by SemVer.

Major version releases will not guarantee backwards compatibility.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.3.0] - 2023-11-19

### Added
- Allow setting timezone on Adapter instantiation.
- Update LICENSE file copyright year.
- Setup static analysis.
- Update README.md


## [1.2.1] - 2023-06-19

### Fixed
- Financial Year type validation on getting business week.


## [1.2.0] - 2023-04-23

### Added
- PHP 8 Compatibility.
- CHANGELOG.md


## [1.1.1] - 2019-09-23

### Added
- Instantiate financial year object with start date, financial year type & whether it should use 52/53 weeks.
- Allow setting financial year start date and trigger recalculation of end date.
- Allow setting financial year weeks and trigger recalculation of end date.


src/DateTimeAdapter.php
<?php

namespace RoussKS\FinancialYear;

use DateInterval;
use DatePeriod;
use DateTime;
use DateTimeImmutable;
use DateTimeInterface;
use DateTimeZone;
use RoussKS\FinancialYear\Exceptions\ConfigException;
use RoussKS\FinancialYear\Exceptions\Exception;
use Traversable;

/**
 * Implementation of PHP DateTime FinancialYear Adapter
 *
 * Class DateTimeAdapter
 *
 * @package RoussKS\FinancialYear
 */
class DateTimeAdapter extends AbstractAdapter implements AdapterInterface
{
    /**
     * @var DateTimeImmutable
     */
    protected $fyStartDate;

    /**
     * @var DateTimeImmutable
     */
    protected $fyEndDate;

    /**
     * @var DateTimeZone|null
     */
    private $dateTimeZone;

    /**
     * DateTimeAdapter constructor.
     *
     * @param string $fyType
     * @param DateTime|DateTimeImmutable|DateTimeInterface|string $fyStartDate // string must be of ISO-8601 format 'YYYY-MM-DD'
     * @param bool $fiftyThreeWeeks
     * @param DateTimeZone|string|null $dateTimeZone // this will be used only and only if a string was provided for start date
     *
     * @return void
     *
     * @throws ConfigException
     * @throws Exception
     */
    public function __construct(string $fyType, $fyStartDate, bool $fiftyThreeWeeks = false, $dateTimeZone = null)
    {
        parent::__construct($fyType, $fiftyThreeWeeks);

        // First set the timezone, then the start date and then auto calculate the end date of the financial year.
        $this->setDateTimeZone($dateTimeZone);
        $this->setFyStartDate($fyStartDate);

        $this->autoSetFyEndDateByStartDate();
    }

    /**
     * {@inheritdoc}
     *
     * Extend parent class in order to recalculate end date if the business year weeks change.
     *
     * @throws Exception
     */
    public function setFyWeeks(bool $fiftyThreeWeeks = false): void
    {
        $originalFyWeeks = $this->fyWeeks;

        parent::setFyWeeks($fiftyThreeWeeks);

        // Reset the financial year's end date according to the weeks setting.
        if ($originalFyWeeks !== null && $originalFyWeeks !== $this->fyWeeks) {
            $this->autoSetFyEndDateByStartDate();
        }
    }

    /**
     * {@inheritdoc}
     *
     * @return DateTimeImmutable
     */
    public function getFyStartDate(): DateTimeInterface
    {
        return $this->fyStartDate;
    }

    /**
     * {@inheritdoc}
     *
     * @param  DateTime|DateTimeImmutable|string $date
     *
     * @throws Exception
     */
    public function setFyStartDate($date): void
    {
        // fyStartDate property is an immutable object.
        $originalFyStartDate = $this->fyStartDate;

        $this->fyStartDate = $this->getDateObject($date);

        $this->validateStartDate();

        // If this method execution is not triggered on instantiation (constructor) which performs the same action,
        // recalculate financial year's end date from current settings,
        // even if the new start date is the same as the previous one (why re-setting the same date?).
        if ($originalFyStartDate !== null) {
            $this->autoSetFyEndDateByStartDate();
        }
    }

    /**
     * {@inheritdoc}
     *
     * @return DateTimeImmutable
     */
    public function getFyEndDate(): DateTimeInterface
    {
        return $this->fyEndDate;
    }

    /**
     * {@inheritdoc}
     *
     * First check for calendar type and the return the corresponding value.
     * Otherwise, it is business type as the only other available.
     *
     * @return DatePeriod|DateTimeImmutable[]
     *
     * @throws Exception
     * @throws \Exception
     */
    public function getPeriodById(int $id): Traversable
    {
        return new DatePeriod(
            $this->getFirstDateOfPeriodById($id),
            DateInterval::createFromDateString('1 day'),
            $this->getLastDateOfPeriodById($id)
        );
    }

    /**
     * {@inheritdoc}
     *
     * @return DatePeriod|DateTimeImmutable[]
     *
     * @throws Exception
     */
    public function getBusinessWeekById(int $id): Traversable
    {
        return new DatePeriod(
            $this->getFirstDateOfBusinessWeekById($id),
            DateInterval::createFromDateString('1 day'),
            $this->getLastDateOfBusinessWeekById($id)
        );
    }

    /**
     * {@inheritdoc}
     *
     * @param  DateTime|DateTimeImmutable|string $date
     *
     * @throws Exception
     */
    public function getPeriodIdByDate($date): int
    {
        $dateTime = $this->getDateObject($date);

        $this->validateDateBelongsToCurrentFinancialYear($dateTime);

        for ($id = 1; $id <= $this->fyPeriods; $id++) {
            // If the date is between the start and the end date of the period, get the period's id.
            if ($dateTime >= $this->getFirstDateOfPeriodById($id) && $dateTime <= $this->getLastDateOfPeriodById($id)) {
                return $id;
            }
        }

        // We can never reach this stage.
        // However, added for keeping the IDEs happy of non returned value.
        throw new Exception('A period could not be found for the requested date.');
    }

    /**
     * {@inheritdoc}
     *
     * @param  DateTime|DateTimeImmutable|string $date
     *
     * @throws Exception
     */
    public function getBusinessWeekIdIdByDate($date): int
    {
        $dateTime = $this->getDateObject($date);

        if (!$this->isBusinessType($this->getType())) {
            throw new ConfigException('Business weeks are set only for a business type financial year.');
        }

        $this->validateDateBelongsToCurrentFinancialYear($dateTime);

        for ($id = 1; $id <= $this->fyWeeks; $id++) {
            if (
                $dateTime >= $this->getFirstDateOfBusinessWeekById($id) &&
                $dateTime <= $this->getLastDateOfBusinessWeekById($id)
            ) {
                return $id;
            }
        }

        // We can never reach this stage.
        // However, added for keeping the IDEs happy of non returned value.
        throw new Exception('A business week could not be found for the specified date.');
    }

    /**
     * {@inheritdoc}
     *
     * First check for calendar type.
     * Otherwise, it will be business type as no other is supported.
     *
     * @return DateTimeImmutable
     *
     * @throws Exception
     */
    public function getFirstDateOfPeriodById(int $id): DateTimeInterface
    {
        $this->validateConfiguration();

        $this->validatePeriodId($id);

        // If 1st period, get the start of the financial year, regardless of the type.
        if ($id === 1) {
            return $this->getFyStartDate();
        }

        // In calendar type, fyPeriods are always 12 as the months,
        // regardless of the start date within the month.
        if ($this->isCalendarType($this->getType())) {
            return $this->getFyStartDate()->modify('+' . ($id - 1) . ' months');
        }

        // Otherwise return business type calculation.
        return $this->getFyStartDate()->modify('+' . ($id - 1) * 4 . ' weeks');
    }

    /**
     * {@inheritdoc}
     *
     * First check for calendar type.
     * Otherwise, it will be business type as no other is supported.
     *
     * @return DateTimeImmutable
     *
     * @throws Exception
     */
    public function getLastDateOfPeriodById(int $id): DateTimeInterface
    {
        $this->validateConfiguration();

        $this->validatePeriodId($id);

        // If last period, get the end of the financial year, regardless of the type.
        if ($id === $this->fyPeriods) {
            return $this->getFyEndDate();
        }

        // In calendar type, fyPeriods are always 12 as the months,
        // regardless of the start date within the month.
        if ($this->isCalendarType($this->getType())) {
            // Otherwise calculate for business type.
            return $this->getFyStartDate()->modify('+' . $id . ' months')->modify('-1 day');
        }

        // Otherwise calculate for business type.
        return $this->getFyStartDate()->modify('+' . $id * 4 . ' weeks')->modify('-1 day');
    }

    /**
     * {@inheritdoc}
     *
     * @return DateTimeImmutable
     *
     * @throws Exception
     */
    public function getFirstDateOfBusinessWeekById(int $id): DateTimeInterface
    {
        $this->validateConfiguration();

        $this->validateBusinessWeekId($id);

        // If 1st week, get the start of the financial year.
        if ($id === 1) {
            return $this->getFyStartDate();
        }

        return $this->getFyStartDate()->modify('+' . ($id - 1) . ' weeks');
    }

    /**
     * {@inheritdoc}
     *
     * @return DateTimeImmutable
     *
     * @throws Exception
     */
    public function getLastDateOfBusinessWeekById(int $id): DateTimeInterface
    {
        $this->validateConfiguration();

        $this->validateBusinessWeekId($id);

        // If last week, get the end of the financial year.
        if ($id === $this->fyWeeks) {
            return $this->getFyEndDate();
        }

        return $this->getFyStartDate()->modify('+' . $id . ' weeks')->modify('-1 day');
    }

    /**
     * {@inheritdoc}
     *
     * @return DatePeriod|DateTimeImmutable[]
     *
     * @throws Exception
     */
    public function getFirstBusinessWeekByPeriodId(int $id): Traversable
    {
        return $this->getBusinessWeekById(($id - 1) * 4 + 1);
    }

    /**
     * {@inheritdoc}
     *
     * @return DatePeriod|DateTimeImmutable[]
     *
     * @throws Exception
     */
    public function getSecondBusinessWeekByPeriodId(int $id): Traversable
    {
        return $this->getBusinessWeekById(($id - 1) * 4 + 2);
    }

    /**
     * {@inheritdoc}
     *
     * @return DatePeriod|DateTimeImmutable[]
     *
     * @throws Exception
     */
    public function getThirdBusinessWeekOfPeriodId(int $id): Traversable
    {
        return $this->getBusinessWeekById(($id - 1) * 4 + 3);
    }

    /**
     * {@inheritdoc}
     *
     * @return DatePeriod|DateTimeImmutable[]
     *
     * @throws Exception
     */
    public function getFourthBusinessWeekByPeriodId(int $id): Traversable
    {
        return $this->getBusinessWeekById($id * 4);
    }

    /**
     * {@inheritdoc}
     *
     * @return DatePeriod|DateTimeImmutable[]
     *
     * @throws Exception
     */
    public function getFiftyThirdBusinessWeek(): Traversable
    {
        return $this->getBusinessWeekById(53);
    }

    /**
     * First check if calendar type and return value.
     * If not calendar type, it is business type (as the only other option available supported and always set).
     * So we can safely return the relevant value.
     *
     * {@inheritdoc}
     *
     * @return DateTimeImmutable
     */
    public function getNextFyStartDate(): DateTimeInterface
    {
        // For calendar type, the next year's start date is + 1 year.
        if ($this->isCalendarType($this->getType())) {
            return $this->getFyStartDate()->modify('+1 year');
        }

        // For business type, the next year's start date is + number of weeks.
        // As a financial year would have 52 or 53 weeks, the param handles it.
        return $this->getFyStartDate()->modify('+' . $this->fyWeeks . ' weeks');
    }

    /**
     * Validate that the start date is not disallowed.
     *
     * @return void
     *
     * @throws ConfigException
     */
    protected function validateStartDate(): void
    {
        $disallowedFyCalendarTypeDates = ['29', '30', '31'];

        if (
            $this->isCalendarType($this->getType()) &&
            in_array($this->getFyStartDate()->format('d'), $disallowedFyCalendarTypeDates, true)
        ) {
            $this->throwConfigurationException(
                'This library does not support 29, 30, 31 as start dates of a month for calendar type financial year.'
            );
        }
    }

    /**
     * Validate that a date belongs to the set financial year.
     *
     * @param  DateTimeImmutable $dateTime
     *
     * @return void
     *
     * @throws Exception
     */
    protected function validateDateBelongsToCurrentFinancialYear(DateTimeImmutable $dateTime): void
    {
        if ($dateTime < $this->getFyStartDate() || $dateTime > $this->getFyEndDate()) {
            throw new Exception('The requested date is out of range of the current financial year.');
        }
    }

    /**
     * Automatically set the financial year's end date by the current start date.
     *
     * We will set end date from the start date object which should be present.
     * Both types calculate end date relative to next financial year start date.
     * As that is automatically calculated for us, regardless of type, we just subtract 1 day.
     *
     * @return void
     */
    protected function autoSetFyEndDateByStartDate(): void
    {
        $this->fyEndDate = $this->getNextFyStartDate()->modify('-1 day');
    }

    /**
     * Get the DateTimeZone currently set
     *
     * @return DateTimeZone|null
     */
    protected function getDateTimeZone(): ?DateTimeZone
    {
        return $this->dateTimeZone;
    }

    /**
     * @param DateTimeZone|string|null $dateTimeZone
     * @return void
     * @throws ConfigException
     */
    protected function setDateTimeZone($dateTimeZone = null): void
    {
        if ($dateTimeZone === null) {
            return;
        }

        if ($dateTimeZone instanceof DateTimeZone) {
            $this->dateTimeZone = $dateTimeZone;
            return;
        }

        if (is_string($dateTimeZone)) {
            try {
                $this->dateTimeZone = new DateTimeZone($dateTimeZone);
                return;
            } catch (\Exception $ex) {
                // Catch exception, set null timezone string and throw config exception.
                $this->throwConfigurationException('Invalid dateTimeZone string: ' . $dateTimeZone);
            }
        }

        $this->throwConfigurationException('Invalid dateTimeZone parameter');
    }

    /**
     * Get & validate a DateTimeImmutable object for the given parameter.
     * If the object is generated, we set it to the start of the day (0, 0) with setTime.
     * setTime will not return false for valid input of hours and minutes.
     *
     * @param  DateTime|DateTimeImmutable|string $date
     *
     * @return DateTimeImmutable
     *
     * @throws Exception
     */
    protected function getDateObject($date): DateTimeImmutable
    {
        $dateTime = $this->generateDateTimeImmutableObject($date);

        // Validation that the datetime object was created and set to the start of the day.
        if (!$dateTime) {
            throw new Exception(
                'Invalid date format. Not a valid ISO-8601 date string or DateTime/DateTimeImmutable object.'
            );
        }

        /** @var DateTimeImmutable $dateTime */
        $dateTime = $dateTime->setTime(0, 0);

        // We have set a valid hour and minutes, so false is not a possible result of the above method.
        return $dateTime;
    }

    /**
     * Generate and return a DateTimeImmutable object for the given $date parameter.
     *
     * First check if we have received an object relevant to the adapter and return it.
     * This can be either a DateTime or DateTimeImmutable object.
     *
     * Otherwise, create the object regardless of the type with createFromFormat.
     * It will return false if it fails.
     *
     * @param  DateTime|DateTimeImmutable|string $date
     *
     * @return DateTimeImmutable|false
     */
    protected function generateDateTimeImmutableObject($date)
    {
        if ($date instanceof DateTime) {
            return DateTimeImmutable::createFromMutable($date);
        }

        if ($date instanceof DateTimeImmutable) {
            return $date;
        }

        return DateTimeImmutable::createFromFormat('Y-m-d', $date, $this->getDateTimeZone());
    }
}
